File: odbc_tests/tests/e2e/put_get/put_get_sse_stage.cpp
#include <filesystem>
#include <fstream>
#include <string>

#include <catch2/catch_test_macros.hpp>

#include "Connection.hpp"
#include "get_data.hpp"
#include "put_get_utils.hpp"
#include "utils.hpp"

namespace fs = std::filesystem;
using namespace pg_utils;

static std::string create_sse_stage(Connection& conn, const std::string& stage_name) {
  conn.execute("CREATE TEMPORARY STAGE IF NOT EXISTS " + stage_name + " ENCRYPTION = (TYPE = 'SNOWFLAKE_SSE')");
  return stage_name;
}

TEST_CASE("should put and get file on SSE stage", "[put_get]") {
  // Given Stage with server-side encryption (SNOWFLAKE_SSE)
  Connection conn;
  const std::string stage = create_sse_stage(conn, unique_stage_name("ODBCTST_SSE_PUT_GET"));

  TempTestDir upload_dir("odbc_sse_upload_");
  fs::path test_file = write_text_file(upload_dir.path(), "sse_test.txt", "hello sse\n");

  // When File is uploaded using PUT command
  std::string put_sql = "PUT 'file://" + as_file_uri(test_file) + "' @" + stage + " AUTO_COMPRESS=FALSE OVERWRITE=TRUE";
  auto put_stmt = conn.execute_fetch(put_sql);

  // Then File should be uploaded successfully
  CHECK(get_data<SQL_C_CHAR>(put_stmt, PUT_ROW_STATUS_IDX) == "UPLOADED");

  // When File is downloaded using GET command
  TempTestDir download_dir("odbc_sse_get_");
  std::string get_sql = "GET @" + stage + "/sse_test.txt 'file://" + as_file_uri(download_dir.path()) + "/'";
  auto get_stmt = conn.execute_fetch(get_sql);

  // Then File should be downloaded
  CHECK(get_data<SQL_C_CHAR>(get_stmt, GET_ROW_STATUS_IDX) == "DOWNLOADED");

  // And Have correct content
  fs::path downloaded = download_dir.path() / "sse_test.txt";
  REQUIRE(fs::exists(downloaded));
  std::ifstream ifs(downloaded);
  std::string content((std::istreambuf_iterator<char>(ifs)), std::istreambuf_iterator<char>());
  CHECK(content == "hello sse\n");
}

TEST_CASE("should put and get file on SSE stage with DIRECTORY enabled", "[put_get]") {
  // Given Stage with server-side encryption and DIRECTORY enabled
  Connection conn;
  const std::string stage = unique_stage_name("ODBCTST_SSE_DIR");
  conn.execute("CREATE TEMPORARY STAGE IF NOT EXISTS " + stage +
               " ENCRYPTION = (TYPE = 'SNOWFLAKE_SSE') DIRECTORY = (ENABLE = TRUE)");

  TempTestDir upload_dir("odbc_sse_dir_");
  fs::path test_file = write_text_file(upload_dir.path(), "test.txt", "Initial contents\n");

  // When File is uploaded using PUT command
  std::string put_sql = "PUT 'file://" + as_file_uri(test_file) + "' @" + stage + " AUTO_COMPRESS=FALSE OVERWRITE=TRUE";
  auto put_stmt = conn.execute_fetch(put_sql);

  // Then File should be uploaded successfully
  CHECK(get_data<SQL_C_CHAR>(put_stmt, PUT_ROW_STATUS_IDX) == "UPLOADED");

  // When File is downloaded using GET command
  TempTestDir download_dir("odbc_sse_dir_get_");
  std::string get_sql = "GET @" + stage + "/test.txt 'file://" + as_file_uri(download_dir.path()) + "/'";
  auto get_stmt = conn.execute_fetch(get_sql);

  // Then File should be downloaded
  CHECK(get_data<SQL_C_CHAR>(get_stmt, GET_ROW_STATUS_IDX) == "DOWNLOADED");

  // And Have correct content
  fs::path downloaded = download_dir.path() / "test.txt";
  REQUIRE(fs::exists(downloaded));
  std::ifstream ifs(downloaded);
  std::string content((std::istreambuf_iterator<char>(ifs)), std::istreambuf_iterator<char>());
  CHECK(content == "Initial contents\n");
}

File: sf_core/tests/e2e/put_get/put_get_sse_stage.rs
use crate::common::file_utils::{create_test_file, path_to_sql_uri};
use crate::common::put_get_common::{GetResult, PutResult};
use crate::common::snowflake_test_client::SnowflakeTestClient;
use uuid::Uuid;

fn random_stage_name(prefix: &str) -> String {
    format!("{}_{}", prefix, Uuid::new_v4().simple())
}

fn create_sse_stage(client: &SnowflakeTestClient, stage_name: &str) {
    client.execute_query(&format!(
        "CREATE TEMPORARY STAGE IF NOT EXISTS {stage_name} ENCRYPTION = (TYPE = 'SNOWFLAKE_SSE')"
    ));
}

#[test]
fn should_put_and_get_file_on_sse_stage() {
    // Given Stage with server-side encryption (SNOWFLAKE_SSE)
    let client = SnowflakeTestClient::connect_with_default_auth();
    let stage_name = random_stage_name("TEST_SSE_PUT_GET");
    create_sse_stage(&client, &stage_name);

    let upload_dir = tempfile::TempDir::new().unwrap();
    let test_file = create_test_file(upload_dir.path(), "sse_test.txt", "hello sse\n");

    // When File is uploaded using PUT command
    let put_sql = format!(
        "PUT 'file://{}' @{stage_name} AUTO_COMPRESS=FALSE OVERWRITE=TRUE",
        path_to_sql_uri(&test_file)
    );
    let result = client.execute_query(&put_sql);

    // Then File should be uploaded successfully
    let mut helper = crate::common::arrow_result_helper::ArrowResultHelper::from_result(result);
    let put: PutResult = helper.fetch_one().expect("Failed to fetch PUT result");
    assert_eq!(put.status, "UPLOADED");

    // When File is downloaded using GET command
    let download_dir = tempfile::TempDir::new().unwrap();
    let get_sql = format!(
        "GET @{stage_name}/sse_test.txt 'file://{}/'",
        path_to_sql_uri(download_dir.path())
    );
    let get_result = client.execute_query(&get_sql);

    // Then File should be downloaded
    let mut helper = crate::common::arrow_result_helper::ArrowResultHelper::from_result(get_result);
    let get: GetResult = helper.fetch_one().expect("Failed to fetch GET result");
    assert_eq!(get.status, "DOWNLOADED");

    // And Have correct content
    let downloaded = download_dir.path().join("sse_test.txt");
    assert!(downloaded.exists(), "Downloaded file should exist");
    let content = std::fs::read_to_string(&downloaded).unwrap();
    assert_eq!(content.trim(), "hello sse");
}

#[test]
fn should_put_and_get_file_on_sse_stage_with_directory_enabled() {
    // Given Stage with server-side encryption and DIRECTORY enabled
    let client = SnowflakeTestClient::connect_with_default_auth();
    let stage_name = random_stage_name("TEST_SSE_DIR");
    client.execute_query(&format!(
        "CREATE TEMPORARY STAGE IF NOT EXISTS {stage_name} \
         ENCRYPTION = (TYPE = 'SNOWFLAKE_SSE') \
         DIRECTORY = (ENABLE = TRUE)"
    ));

    let upload_dir = tempfile::TempDir::new().unwrap();
    let test_file = create_test_file(upload_dir.path(), "test.txt", "Initial contents\n");

    // When File is uploaded using PUT command
    let put_sql = format!(
        "PUT 'file://{}' @{stage_name} AUTO_COMPRESS=FALSE OVERWRITE=TRUE",
        path_to_sql_uri(&test_file)
    );
    let result = client.execute_query(&put_sql);

    // Then File should be uploaded successfully
    let mut helper = crate::common::arrow_result_helper::ArrowResultHelper::from_result(result);
    let put: PutResult = helper.fetch_one().expect("Failed to fetch PUT result");
    assert_eq!(put.status, "UPLOADED");

    // When File is downloaded using GET command
    let download_dir = tempfile::TempDir::new().unwrap();
    let get_sql = format!(
        "GET @{stage_name}/test.txt 'file://{}/'",
        path_to_sql_uri(download_dir.path())
    );
    let get_result = client.execute_query(&get_sql);

    // Then File should be downloaded
    let mut helper = crate::common::arrow_result_helper::ArrowResultHelper::from_result(get_result);
    let get: GetResult = helper.fetch_one().expect("Failed to fetch GET result");
    assert_eq!(get.status, "DOWNLOADED");

    // And Have correct content
    let downloaded = download_dir.path().join("test.txt");
    assert!(downloaded.exists(), "Downloaded file should exist");
    let content = std::fs::read_to_string(&downloaded).unwrap();
    assert_eq!(content.trim(), "Initial contents");
}
